### DISCUSSION TITLE
podman in production: unpriviliged user vs. `--userns=auto`

### DISCUSSION THREAD

**Discussion by tobwen** [containers/podman#13728] [Q&A] [answered]
Based on [this discussion](https://github.com/containers/podman/issues/13706), @rhatdan stated that `podman run --userns=auto` might be a better solution for containers in production launched by podman. For sure, when starting it as a privileged user, you're way more flexible.

But is the isolation in a unique UserNS as secure as running podman from an unprivileged user?

**Comment by rhatdan** [comment, marked as answer]
It is actually more secure.  If you run two containers as a rootless user, they run in the same user namespace so they can attack each other from a User Namespace point of view.

If you run two containers as root with podman run --userns auto, then they run in unigue user namespace and are isolated.

Rootless containers are great for containers run by users on a system, but if you are just running containers on a server, then --userns=auto is a more secure solution. (I plan on writing a blog on this).

**Reply by tobwen** [reply]
Thank you for your explanation. A blog article on this topic would be excellent.

I know numerous people - and I count myself among them - who assumed that containers started by an unprivileged user would be more secure. This is probably related to the fact that an attacker would not have root access if they broke out of the container. The question is to what extent it is possible to break out of the container at all.

**Reply by rhatdan** [reply]
None of the process that run with --userns=auto are running as root and none have any host linux capabiltiies, only User Namespaced capabilities.

**Reply by Luap99** [reply]
Well I can still run --userns=auto as rootless. I just need to configure more uids manually so I would say --userns=auto is more secure than without it.

**Reply by rhatdan** [reply]
Yes but you have much smaller number of UIDs available to run containers in this manner.

**Reply by sbrivio-rh** [reply]
On the networking side, running as root with `--userns=auto` doesn't currently work with any user-mode networking implementation (https://github.com/containers/podman/issues/17840), so, at the moment, you can't enforce strong network isolation (no crafted frames, no ARP spoofing) _by design_, as root.

**Reply by rhatdan** [reply]
Very interesting.
Thanks for pointing this out.  The problem with the document and why I have not published it for a couple of years is that there are pluses and minuses to each way of doing it.

**Reply by vmsh0** [reply]
Just to make sure I understand this right, would the following points be an accurate summary of the discussion?

1. Rootful containers with --userns=auto DO NOT have root capabilities outside the containers, as all container users (including root) are mapped to unprivileged UIDs in the host namespace
2. In addition to point 1, they DO NOT have access to any resources in other containers, as even if they managed to access other containers' filesystem nodes, there would be no UID overlaps and thus no chance for access to restrictively-permissioned files
3. Rootless containers without --userns=auto (i.e. with --userns=host) DO NOT have root capabilites other UIDs' capabilities outside the containers, except for the launching UID and its SUBUIDs
4. However, they DO have capabilities inside other containers with --userns=host, should they manage to mount their filesystems, as they share the same host namespace UIDs
5. Rootless containers with --userns=auto have, essentially, the same exact level of security isolation as rootful containers with --userns=auto, including isolation from and to other containers, as userns-wise they are essentially the same thing (with the main practical difference being what the podman process itself can do in terms of setup, e.g. network, before forking and unsharing)

Point 5 is, imho, especially interesting: why bother, in a production environment, to:
- set up rootless containers, potentially requiring more complicate networking setup with an helper daemon (-> if you run podman as root, it can actually set up the stuff you need)
- set up images so that you run stuff inside as non-root users (-> when both container root and container non-root are mapped to non-root UIDs on the host, it is essentially the same as a rootless container)

So --userns=auto looks to me like the best of both worlds: podman is run with appropriate privileges until needed, and then the container essentially becomes rootless when started. The only limitation, is that normal users cannot start containers, which is then the only real use case for rootless podman, having at the end of the day nothing to do with container isolation. Am I getting this right?

**Reply by pomology** [reply]
@vmsh0, insightful points.  I hope @rhatdan can comment on your summery.  I'd like to know if he would agree with these points.

**Reply by vmsh0** [reply]
FWIW, after posting that I have taken the time to study `user_namespaces(7)` and while the matter is slightly more complex than how I framed it in my post, it looks like the main takeaway is essentially correct: the mere fact that a separate user namespace is created and that inner root is mapped to outer non-root achieves the isolation, not so much what user you use to create the namespace itself (which however, beware, would be relevant in other use cases).

This is, broadly, for the following reasons:
- filesystem access happens through mapped UIDs and GIDs
- privileged operations on filesystems can only be performed by UID 0 of the user namespace the filesystem was mounted in, or outer namespaces
- capabilities are scoped to the user namespace
- all the other kind of namespaces are scoped to the user namespace they were created from
- as a final catchall, global state (e.g. system time, block devices) can only be managed in the initial namespace

**Reply by rhatdan** [reply]
Yes I agree with all of these points.  The only difference between rootful podman --userns=auto and rootless podman --userns=auto, is what the `podman` command itself can do, or be tricked to be done.

In a rootful podman could be tricked while pulling and writing the image to write content in a location outside of container storage.  (We use features of the OS to prevent this, like chroot and enhanced security.) These risks can also be mitigated by running only images that are trust worthy. (Don't run random crap from the internet on your system.)  In the case of a rootless podman, the podman process can do everything that rootless user can do, so the attack would be limited to the user access.

I believe for a server running all containers in a separate User Namespace with isolated UIDs is the most secure way to run workloads.

**Reply by tobwen** [reply]
Thanks for the information!

Can we put this important information in a more prominent place? Maybe as a markdown table or something like that in the wiki or a readme?

**Reply by sbrivio-rh** [reply]
> Rootless containers with --userns=auto have, essentially, the same exact level of security isolation as rootful containers with --userns=auto, including isolation from and to other containers, as userns-wise they are essentially the same thing (with the main practical difference being what the podman process itself can do in terms of setup, e.g. network, before forking and unsharing)

Not so much from my (mostly networking-driven) perspective. A rootful container:
- has a network interface set up by root, which can receive and send pretty much any vaguely Ethernet-resembling frame
- is started as root, so if you manage to get arbitrary execution in Podman before it switches to a new namespace you're root too (but that's the part you mentioned already)

Conversely, a rootless container:
- has no ("outer") network interface, and that's [enforced](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/net/core/rtnetlink.c?id=ae90f6a6170d7a7a1aa4fddf664fbd093e3023bc#n3509) by the kernel
- isn't started as root... obvious dual of the second point above

**Reply by vmsh0** [reply]
> Not so much from my (mostly networking-driven) perspective.

That is exactly what has driven me to investigate the problem. I have much dislike for working with the various userspace networking options. :)

**Reply by pomology** [reply]
Same here. There are complex container scenarios I've encountered where rootless networking can cause tremendously convoluted issues (though in other situations it's been no big deal...) 

Starting the container as root should be safer though if you have SELinux enabled, right? (Besides using only trusted, vetted images...)

@rhatdan, if SELinux is enabled, what's your take on the networking risks that @sbrivio-rh raised?

**Reply by sbrivio-rh** [reply]
Eh? SELinux doesn't filter packets.

**Reply by pomology** [reply]
@sbrivio-rh: in a rootfull podman networking context, what is the worst that malicious packets can do if SELinux is enabled and correctly configured?

**Reply by tobwen** [reply]
Don't put all of your trust into SELinux when you're outside of RedHat. Seems like many Linux distributions have issues with it.

Article: https://klecko.github.io/posts/selinux-bypasses/
Discussion: https://news.ycombinator.com/item?id=41946453

**Reply by pomology** [reply]
Thanks for the links. Very interesting.  I've only ever used SELinux with RHEL, but good to know to watch for issues with other distros.

**Reply by sbrivio-rh** [reply]
> Don't put all of your trust into SELinux when you're outside of RedHat. Seems like many Linux distributions have issues with it.
> Article: https://klecko.github.io/posts/selinux-bypasses/

...which explains how to _disable_ SELinux on some Android devices. :confused:

> @sbrivio-rh: in a rootfull podman networking context, what is the worst that malicious packets can do if SELinux is enabled and correctly configured?

Exactly the same as malicious packets can do without Podman, or without SELinux. You can spoof ARP, craft semi-random things to crash/DoS routers, etc. Network isolation is still there at Layer-2, but all you need to do is to get control over the network interface in the container to send out whatever you want. I covered that quickly in slide 6 of this [deck](https://static.sched.com/hosted_files/devconfcz2023/b5/pasta_devconf.pdf) for this [presentation](https://www.youtube.com/watch?v=tlxDmUPc4WY).

**Reply by Luap99** [reply]
But doesn't spoofing and crafting require CAP_NET_RAW? This is not given to a container by default and I see very little reason to do so and if an application really needs this they most likely cannot use pasta/slirp4netns anyway unless they only use it for local connections inside the netns.

**Reply by tobwen** [reply]
> ...which explains how to _disable_ SELinux on some Android devices. 😕

I knew the comment was coming, so I copied the discussion page 😃. Some users say that SELinux unfortunately isn't always implemented reliably.

**Reply by sbrivio-rh** [reply]
> But doesn't spoofing and crafting require CAP_NET_RAW? This is not given to a container by default

Sure, but if you start the container as root, you rely on Podman to drop it (and here I'm assuming a bug in Podman). Or one could also pass `--privileged` for whatever other reason and miss this aspect.

On top of that, what is sent **to** the container namespace (not necessarily to processes running into it) is not as restricted.

> I knew the comment was coming, so I copied the discussion page 😃. Some users say that SELinux unfortunately isn't always implemented reliably.

Yeah, I read that as well, but... it starts with a mention of SELinux on Debian as an experiment. Now, sure, some people spent [effort](https://wiki.debian.org/SELinux/Setup) to get that somehow working, but the "native" LSM on Debian is AppArmor. It's not so much of a matter of SELinux _implementation_, it's rather about choices, configuration and integration.

**Reply by rhatdan** [reply]
SELinux by default implements unconfined networking, we rely on network namespace and dropped capabilities to protect the network.

If you start with the assumption that Podman is broken then running containers as root is a risk.  Same as with the Kernel is broken or systemd  or ...

SELinux confines activity within the container, especially about the file system, and using kernel file systems.

**Comment by tobwen** [comment]
Can you please remember to write a short blog post stating that a rootful _podman_ is safer than a rootless _podman_? I'm really struggling to convince other users of this. Maybe a short notice in the README would be enough?

**Reply by Luap99** [reply]
Well this sentence is not true. We explicitly talk about --userns auto is more secure. This is a very important difference.
The secure part is that with userns auto all containers run in a different user namespace (different uids on the host). You can achieve the same as rootless user so I wouldn't say rootful is more secure than rootless.

**Reply by tobwen** [reply]
Sure, `--userns auto` is mandatory. However, I assumed that there are other security functions in the kernel that could currently only be accessed by root. It is wonderful that this has now been clarified.

**Reply by rhatdan** [reply]
I have a blog in progress, the problem is it is arguable which is more secure.
$ sudo podman run --userns=auto ...  Runs each containers in a separate user namespace where each container is running with different UIDs and almost assuradly does not match any other UIDs on the host. BUT the podman command runs as root which means it pulls images as root.  Pulling images as root can be risky, because flaws in the pulling could (have in the past) cause root to write anywhere.
$ podman run ... Runs the container in a single user namespace where it could attack the users homedir and potentially get a users data.  Also all containers run in the same user namespace.

**Reply by tobwen** [reply]
> $ podman run ... Runs the container in a single user namespace where it could attack the users homedir and potentially get a users data. Also all containers run in the same user namespace.

Would this also be the case if you run each container under a different user? Or would a different namespace then be assigned?

**Reply by rhatdan** [reply]
If you run each container under a different user and managed the /etc/subuid and /etc/subgid files to maintain separation, this would be safer, although they could access content in user homedir.

$ podman run --userns=nomap ...

Would solve this problem.

But this is way more complex.  You could also setup a Huge /etc/subuid range for a user and then run lots of containers for that user with --userns=auto ...  Which is probably the most secure.

**Comment by cgwalters** [comment]
One thing related to this...what would I think be quite cool is integration with systemd `DynamicUser=yes` at least for quadlet containers (we'd still need to mount the image as privileged, so doing this would probably require some lower level systemd integration).

**Reply by cgwalters** [reply]
Yeah, the `containers` user here is IMO pretty suboptimal.  Is something actually intended to create that user today automatically?

**Comment by pomology** [comment]
@rhatdan, were you ever able to write the blog post above regarding the security implications of rootfull podman as potentially more secure vs rootless?  I'd love to read it if you can share a link.  Thank you!

**Reply by rhatdan** [reply]
I have not, although your request spurred me to revisit and do a bunch of rewriting.  The original write was in 2022, and does not cover Quadlet or idmap file systems.  I will add these and see if we can finally get around to publishing it.

**Reply by rhatdan** [reply]
@pomology Send me and email and I can probably preview what I have now to you.

**Reply by pomology** [reply]
@rhatdan Thank you! I sent you an email just now.

**Reply by tobwen** [reply]
I'm also interested. Do a MEAP, I'm a legit owner of https://www.manning.com/books/podman-in-action

**Reply by rhatdan** [reply]
Send me an email, and I will send the preliminary doc to you.

**Reply by kavishgr** [reply]
I'm also interested. I sent you an email.

**Comment by rdbisme** [comment]
Hello, what's the best practice here nowadays? 

I'm currently deploying containerized services creating a dedicated non-root user on my system per "pod". Enable lingering and running the container within that dedicated user in rootless mode. 

How does it look? Would it be better to run them instead with `root` + `--userns=auto`?

**Reply by giuseppe** [reply]
from a security PoV at runtime they behave in the same way.  The main difference of using `--userns=auto` compared to a different system user per pod is that the code for pulling the image and setting up the container is also running in a user namespace.  So the "user per pod" setup is slightly safer, than running as root with `--userns=`auto`.  The cost is also higher though, since you are not able to share storage among different users

**Comment by videoMonkey** [comment]
I was very excited at the thought of not creating a user per pod but I am running into a problem if I run `sudo podman run -d --name vaultwarden --userns=auto -p 8000:80 docker.io/vaultwarden/server:latest` I get the following feedback
```
ERRO[0000] Cannot find mappings for user "containers": no subuid ranges found for user "containers" in /etc/subuid
Error: creating container storage: not enough unused IDs in user namespace
```
and maybe I need to read more about how this all works, but I tried setting subuid's and subgid's for the user 'container', that did nothing, I tried doing it for root and then realized root already had a lot. I tried running the command without sudo, that worked.

I was hoping to have an uncomplicated and secure way to have multiple admins be able to tinker with containers and this seemed to be that promise. If i could get it going....

**Reply by sbrivio-rh** [reply]
Could that be https://bugzilla.redhat.com/show_bug.cgi?id=2382662?

**Reply by vmsh0** [reply]
> ```
> ERRO[0000] Cannot find mappings for user "containers": no subuid ranges found for user "containers" in /etc/subuid
> Error: creating container storage: not enough unused IDs in user namespace
> ```
What you're saying is correct ("I was hoping to have an uncomplicated and secure way to have multiple admins be able to tinker with containers and this seemed to be that promise.").

But this functionality requires some initial setup in the operating system for the users running the containers. You should read `subuid(5)` and set up proper mappings, as reported by the error message. This can also be easily integrated in user creation (and it is, in some distros.)

**Reply by vmsh0** [reply]
I.e., "auto" does not mean that the system configures itself automatically. "auto" means that a suitable ID range is selected from the available range.

**Reply by jzcad1828** [reply]
I think I also ran across this issue. I added the example from this documentation to my gid files and it fixed it for me:
 https://docs.podman.io/en/v4.6.1/markdown/options/userns.container.html:

The --userns=auto flag `requires that the user name containers be specified in the /etc/subuid and /etc/subgid files`, with an unused range of subordinate user IDs that Podman containers are allowed to allocate. See subuid(5).

Example: containers:2147483647:2147483648

**Comment by pierrecnalb** [comment]
I am a bit surprised by the conclusion of this discussion:
why is the recommendation for servers to use rootful containers with `--userns=auto`, instead of using a dedicated rootless user with `--userns=auto`?
Basically, instead of adding the `containers:2147483647:2147483648` to `/etc/subuid` and `/etc/subgid`, why not adding a real dedicated user that would run all containers in rootless mode?
Something like:
```sh
$ cat /etc/subuid
dwalsh:100000:65536
rootless:2147483647:2147483648
$ cat /etc/subgid
dwalsh:100000:65536
rootless:2147483647:2147483648
```
where `rootless` would be a real user and all containers and  pods would be running with that user using `--userns=auto`

Any comment on that approach @rhatdan, @giuseppe ?

**Reply by giuseppe** [reply]
what do you mean a dedicated user?

If you want to run containers as that user, you can already do that, if you login with the `rootless` user you can already create containers with `--userns=auto`.  You don't need to be root

**Reply by pierrecnalb** [reply]
Sure! By dedicated user, I meant using a single `rootless` user for managing and running all the containers/pods, rather than having one rootless user per pod/container for example. Basically doing the exact same things as root (userns=auto + extended range in subuid/subgid), except with a user so that it can be rootless.

To ask differently, this whole thread is about the most secure way to run podman in production. The [accepted answer](https://github.com/containers/podman/discussions/13728#discussioncomment-2477524) advised to use root + --userns=auto because multiple rootless containers are running in the same user namespace, therefore they can attach each other.
So my question is, why isn't it advised to use rootless + userns=auto instead of rootful?

**Reply by giuseppe** [reply]
rootful with `--userns=auto` is just easier to manage.  If you get it to work with rootless, that is better, because the container runtime itself is confined in a user namespace

**Reply by pierrecnalb** [reply]
Many thanks for the clarification!
By "easier to manage", is that only because root already exist as a user? or is that due to some complex differences with how rootless works (e.g. the pasta network stack etc?)?
In other words, are there any known big difficulties that would prevent it from being the recommended approach for production servers?

**Reply by giuseppe** [reply]
you need to make sure the user session is always kept alive (you can set lingering for that user), otherwise containers are also killed when the user session is terminated.  With root you don't have this problem

**Reply by pierrecnalb** [reply]
Got it! Thanks!
Could you confirm though, that for rootless, it is enough to increase the range of uid/guid for the dedicated user in  `/etc/subuid` and `/etc/subgid` and the `--userns=auto` option would work similarly as root with its `containers` entry?
That is, instead of the following for rootful containers,
```sh
$ cat /etc/subuid
me:100000:65536
containers:2147483647:2147483648
$ cat /etc/subgid
me:100000:65536
containers:2147483647:2147483648
```
I would have the following for rootless containers running with `rootless` user:
```sh
$ cat /etc/subuid
me:100000:65536
rootless:2147483647:2147483648
$ cat /etc/subgid
me:100000:65536
rootless:2147483647:2147483648
```

Or is there anything else to be aware of?

**Reply by giuseppe** [reply]
yes, that is enough.

Of course the usual rootless limitations apply, especially when it comes to networks

**Reply by sbrivio-rh** [reply]
Beware of https://github.com/containers/podman/issues/17840 as well. But as far as I understand there's a workaround in place for the moment that should make things work.

**Reply by dbie0** [reply]
To summarise already mentioned options:
1. Run all containers as root + --userns=auto - helps mitigate container to container attacks, but container escape consequences are bad
2. Run all containers as one specific dedicated user with extended subuid + --userns=auto - helps mitigate container to container attacks, limits container escape consequences

The issue I see with option 2 is: when container escape happens, attacker will have access to all other containers. Therefore maybe option 3 is a way to go?

3. Run each container as it's own dedicated user

It seems totally daunting to manage though. Maybe to be used for containers requiring stronger isolation.

**Comment by videoMonkey** [comment]
My struggle was with the advice pointed to in this thread. To make a user and attach a container to it this is what I've done, vaultwarden is an example of what I am trying to do.
This has worked just fine for me. 

`sudo useradd vaultwarden`
`sudo loginctl enable-linger vaultwarden`
`mkdir -p /home/vaultwarden/vaultwarden_data`
`mkdir -p /home/vaultwarden/.config/containers/systemd`
`sudo chown -R vaultwarden:vaultwarden /home/vaultwarden/`
to make sure the following command works `sudo dnf install systemd-container`
`sudo machinectl shell vaultwarden@` or potentially `sudo -i -u vaultwarden` I'm not sure what the ups and downs are

```
cat < ~/.config/containers/systemd/vaultwarden.container 
[Unit] 
Description=Vaultwarden Server 
[Container] 
EnvironmentFile=%h/.env 
Image=docker.io/vaultwarden/server:latest 
AutoUpdate=registry 
Network=host 
UserNS=keep-id 
Volume=%h/vaultwarden_data:/data:Z 
Environment=ROCKET_PORT=8000 
[Service] 
# Inform systemd of additional exit status Success
ExitStatus=0 143 
# Extend Timeout to allow time to pull the image 
TimeoutStartSec=900 
[Install] 
# Start by default on boot 
WantedBy=default.target 
# https://docs.podman.io/en/latest/markdown/podman-systemd.unit.5.html 
EOF 
```

`systemctl --user daemon-reload`
`systemctl --user start vaultwarden.service`
`systemctl --user status vaultwarden.service`

This has worked pretty well. But I would like to dedicate a moment to try the tehnique mentioned in this issue, I think it would make management easier.
